qiita_db/processing_job.py

```diff
@@ -337,3 +507,445 @@ def step(self, value):
+        Parameters
+        ----------
+        user : qiita_db.user.User
+            The user creating the workflow
+        root_jobs : list of qiita_db.processing_job.ProcessingJob
+            The root jobs of the workflow
+        name : str, optional
+            The name of the workflow. Default: generated from user's name
+        """
+        with qdb.sql_connection.TRN:
+            # Insert the workflow in the processing_job_worflow table
+            name = name if name else "%s's workflow" % user.info['name']
+            sql = """INSERT INTO qiita.processing_job_workflow (email, name)
+                     VALUES (%s, %s)
+                     RETURNING processing_job_workflow_id"""
+            qdb.sql_connection.TRN.add(sql, [user.email, name])
+            w_id = qdb.sql_connection.TRN.execute_fetchlast()
+            # Connect the workflow with it's initial set of jobs
+            sql = """INSERT INTO qiita.processing_job_workflow_root
+                        (processing_job_workflow_id, processing_job_id)
+                     VALUES (%s, %s)"""
+            sql_args = [[w_id, j.id] for j in root_jobs]
+            qdb.sql_connection.TRN.add(sql, sql_args, many=True)
+            qdb.sql_connection.TRN.execute()
+
+        return cls(w_id)
+
+    @classmethod
+    def from_default_workflow(cls, user, dflt_wf, req_params, name=None):
+        """Creates a new processing workflow from a default workflow
+
+        Parameters
+        ----------
+        user : qiita_db.user.User
+            The user creating the workflow
+        dflt_wf : qiita_db.software.DefaultWorkflow
+            The default workflow
+        req_params : dict of {qdb.software.Command: dict of {str: object}}
+            The required parameters values for the source commands in the
+            workflow, keyed by command. The inner dicts are keyed by
+            parameter name.
+        name : str, optional
+            Name of the workflow. Default: generated from user's name
+
+        Returns
+        -------
+        qiita_db.processing_job.ProcessingWorkflow
+            The newly created workflow
+        """
+        with qdb.sql_connection.TRN:
+            dflt_g = dflt_wf.graph
+
+            # Find the roots of the workflow. That is, the nodes that do not
+            # have a parent in the graph (in_degree = 0)
+            in_degrees = dflt_g.in_degree()
+
+            # We can potentially access this information from the nodes
+            # multiple times, so caching in here
+            all_nodes = {n: (n.command, n.parameters)
+                         for n in in_degrees}
+            roots = {n: (n.command, n.parameters)
+                     for n, d in viewitems(in_degrees) if d == 0}
+
+            # Check that we have all the required parameters
+            root_cmds = set(c for c, _ in viewvalues(roots))
+            if root_cmds != set(req_params):
+                error_msg = ['Provided required parameters do not match the '
```

Comment:
Good catch - I thought I wrote those but looking back at the code I don't see them. Totally not my day yesterday...